jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,486 @@
+        private final int priority;
+        private final long expireNanoTime;
+
+        private Entry(Request request, Response response, Callback callback, int priority)
+        {
+            this.request = request;
+            this.response = response;
+            this.callback = callback;
+            this.priority = priority;
+            Duration maxSuspend = getMaxSuspend();
+            long suspendNanos = NanoTime.now() + maxSuspend.toNanos();
+            if (suspendNanos == Long.MAX_VALUE)
+                --suspendNanos;
+            this.expireNanoTime = maxSuspend.isZero() ? Long.MAX_VALUE : suspendNanos;
+        }
+
+        @Override
+        public long getExpireNanoTime()
+        {
+            return expireNanoTime;
+        }
+
+        private void expire()
+        {
+            // The request timed out, therefore it never acquired a permit.
+            boolean removed = queues.get(priority).remove(this);
+            if (removed)
+            {
+                while (true)
+                {
+                    // See state machine logic in handle() and resume().
+                    int suspended = state.getLo();
+                    if (!state.compareAndSetLo(suspended, suspended - 1))
+                        continue;
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("{} timeout {}", QoSHandler.this, request);
+                    request.setAttribute(EXPIRED_ATTRIBUTE_NAME, true);
+                    failSuspended(request, response, callback, HttpStatus.SERVICE_UNAVAILABLE_503, new TimeoutException());
+                    return;
+                }
+            }
+        }
+
+        private void resume(Callback callback)
+        {
+            try
+            {
+                boolean handled = forward(request, response, callback);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} handled={} {}", QoSHandler.this, handled, request);
+                if (!handled)
+                    failSuspended(request, response, callback, HttpStatus.NOT_FOUND_404, null);
+            }
+            catch (Throwable x)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} failed {}", QoSHandler.this, request, x);
+                failSuspended(request, response, callback, HttpStatus.INTERNAL_SERVER_ERROR_500, x);
+            }
+        }
+    }
+
+    private class Forwarder extends IteratingCallback
+    {
+        private final Queue<Entry> queue = new ConcurrentLinkedQueue<>();
+        private Entry entry;
+
+        private void offer(Entry entry)
+        {
+            queue.offer(entry);
+            iterate();
+        }
+
+        @Override
+        protected Action process()
+        {
+            entry = queue.poll();
+            if (entry == null)
+                return Action.IDLE;
+
+            entry.resume(this);
+            return Action.SCHEDULED;
+        }
```

Comment:
This should not need to be a queue. There should only be either 0 or 1 resumed requests.
```suggestion
        private final AtomicReference<Entry> resumed = new AtomicReference<>();
        private Entry entry;

        private void offer(Entry entry)
        {
            if (!resumed.compareAndSet(null, entry))
                throw new IllegalStateException();
            iterate();
        }

        @Override
        protected Action process()
        {
            entry = resumed.getAndSet(null);
            if (entry == null)
                return Action.IDLE;

            entry.resume(this);
            return Action.SCHEDULED;
        }
```

Reply:
It must be a queue.
If 10 requests are processed by 10 different threads, when they are complete you have 10 threads calling `resume()`, which will find 10 meanwhile suspended requests that are offered to the `Forwarder`.

Perhaps we may be smarter and if the thread is different we don't use the `Forwarder`? 
Otherwise the requests that may be concurrent are serialized.

Reply:
Hmmm now it feels more like a ThreadLocal.
If we have 10 different requests all with their callbacks being completed at the same time, then for each request there are two options:
1. the thread that called `Handler.handle(...)` is the one completing, in that case we only need to take one job from the suspended queue and remember it until we return from handle, when we can call the next `handle`.  Of course this assumes that after completing the callback that the thread returns immediately and does not go blocking on databases trying to commit some result.
2. the thread completing is some other thread not in the scope, in which case we could just make it execute any resumed job itself... or is that a breach of the Callback contract as it may take some time?  Perhaps in that case we need to execute a new thread?

Eitherway, I don't see that we should be taking entries off the suspended queue only to add them to an iteration queue.  Double contention!